### ISSUE TITLE
Compiling the file system RPC interface example results in FlatMessageBuilder buffer too large error.

### ISSUE THREAD

**Issue by zeppenfe** [sandstorm-io/capnproto#405]
When trying to compile the attached services.txt (taken from the Cap'n Proto RPC protocol page), or any other schema file containing a non-empty interface, the capnp compiler errors out with an uncaught exception:

```
$ capnproto-c++-0.5.3/capnp compile -oc++ services.txt
**** Uncaught exception ***
src/capnp/message.c++:232: requirement not met: expected getSegmentsForOutput()[0].end() == array.end(); FlatMessageBuilder's buffer was too large.
stack: 0x7f9da2824fed 0x7f9da28289b7 0x7f9da2a6c681 0x7f9da2a70911 0x7f9da2a70f0d 0x7f9da2a70fde 0x7f9da2a732f1 0x5627c645b125 0x5627c645eabd 0x7f9da2831b2d 0x7f9da282b7e5 0x7f9da2825558 0x7f9da282bb49 0x5627c6441dd2 0x7f9da1f59f45 0x5627c6441e8b
c++: plugin failed: exit code 1
```

[services.txt](https://github.com/sandstorm-io/capnproto/files/726604/services.txt)

**Comment by zeppenfe** [comment]
After finding that compiling the example with the prebuilt capnp compiler provided by the capnproto-dev package (though version 0.4.0), as well as the prebuilt windows binaries (for version 0.5.3) works fine, here are a few more details in the hope that they help you reproduce this problem:

Running Ubuntu 14.04.5 LTS, with g++ (Ubuntu 4.8.4-2ubuntu1~14.04.3) 4.8.4, and following the instructions for installing from a release tarball given [in the manual](https://capnproto.org/install.html#installation-unix). I didn't perform a make install since I do not have root permissions.

Any ideas as to what I could be doing wrong?

**Comment by zeppenfe** [comment]
Digging into this some more, it appears that the installed, 0.4.0 version of the capnpc-c++ compiler is executed by the 0.5.3 version capnp tool. Line 232 in src/capnp/message.c++ only makes sense for version 0.4.0.

Digging into this even more, I discovered that -oc++ executes the capnpc-c++ executable in the PATH, rather than in its own directory. Running with -o./capnpc-c++ fixes the problem.

So yes, admittedly this is a user bug. However, it would be nice to either:
1) Change the default behavior changed to the one that is more likely to work (look for capnpc-c++ in the capnp directory first). What is the use case that makes searching in PATH first desirable?
2) Get a better error message, such as "capnp-c++ version mismatch".

**Comment by kentonv** [comment]
Hmm. I guess the thought was that code generators may come from different packages, and often people want to install different packages to different places. E.g. you might install the Cap'n Proto package from your distro, but then you might want to install a code generator for another language that doesn't have distro packages yet, so you install it from source, and it ends up in a different directory. It's only `capnp` and `capnpc-c++` where this really doesn't make sense, since they come from the same package.

Similarly, it's not completely clear that you'd want to detect "version mismatches" for code generators other than `capnpc-c++` -- it's actually perfectly legitimate to use a new version of `capnp` with an older code generator, or vice versa. The communications protocol between the two is designed to be backwards- and forwards-compatible. Of course, an old code generator might not be able to handle newer language features, but generally it should end up just ignoring the things it doesn't understand...

**Comment by zeppenfe** [comment]
Agreed, it does make sense for different code generators from different packages.

I guess this is a bug in capnpc-c++ then. I would expect a code generator to spit out a warning when it receives something it doesn't understand, and it certainly shouldn't crash with an unhandled exception and unrelated error message.

**Comment by kentonv** [comment]
Yes, the crash is definitely a bug, but I think it is a bug that has already been fixed sometime after 0.4.0. :)

**Comment by kentonv** [comment]
The only reasonable thing I can think of to do here is to print a warning from capnpc-c++ when the compiler is a newer version. Of course, old versions of capnpc-c++ won't have that warning so it will take a while for this change to be effective, but eventually everyone will get a warning when the versions don't match.
